### PR TITLE
Conversion of MessageBody to accept Pin in poll_next 

### PR DESCRIPTION
**Pull request by dunnock**
Starting conversion of `MessageBody` to accept `poll_next(self: Pin<&mut Self...)` as it was adviced in https://github.com/actix/actix-web/issues/1321#issuecomment-579339623 .

This is breaking change and probably should be included in the next version, amended versions accordingly. 

Following versions updated as part of this PR to align with semver:

 `actix-http` 1 -> 2.0.0-alpha

## Checklist:

- [X] Draft implementation of actix-http
- [X] Fix tests in actix-http
- [X] Benchmark & compare with non Pin version
- [X] Prereview 
- [X] Review 

## Move to another PR / tasks
- Publish actix-http with version 2.0.0-alpha to crates
- Fix dependencies (actix-web and awc), update versions:
 `actix-web` 2.0 -> 2.1 (probably should go to 3.0?)
 `awc` 1 -> 2

### CODE REVIEW

**Comment by JohnTitor** [comment]
Could you reword b55aa2e? Otherwise it'll close mentioned issue.

**Comment by dunnock** [comment]
> Could you reword [b55aa2e](https://github.com/actix/actix-web/commit/b55aa2e00cb3406c537a37c2f4124c51e5c817fe)? Otherwise it'll close mentioned issue.

Done, had to squash them but it seems no conflicts so ok

**Comment by dunnock** [comment]
After running benchmarks and separately load testing actix server with ab I do not see any significant difference in performance, pinned version seems working slightly faster but most likely it is just random variance:

## ApacheBenchmark

Tests performed on ubuntu with 32GB RAM and 40 CPU

```
> ab -n 1000000 -c 40 -k ...
```

Version from this branch using Pin in poll_next():

```
> ab -n 1000000 -c 40 -k ...
Concurrency Level:      40
Time taken for tests:   29.673 seconds
Complete requests:      1000000
Failed requests:        0
Keep-Alive requests:    1000000
Total transferred:      220000000 bytes
HTML transferred:       119000000 bytes
Requests per second:    33701.00 [#/sec] (mean)
Time per request:       1.187 [ms] (mean)
Time per request:       0.030 [ms] (mean, across all concurrent requests)
Transfer rate:          7240.45 [Kbytes/sec] received
```

Version from master branch:

```
Concurrency Level:      40
Time taken for tests:   30.945 seconds
Complete requests:      1000000
Failed requests:        0
Keep-Alive requests:    1000000
Total transferred:      220000000 bytes
HTML transferred:       119000000 bytes
Requests per second:    32315.58 [#/sec] (mean)
Time per request:       1.238 [ms] (mean)
Time per request:       0.031 [ms] (mean, across all concurrent requests)
Transfer rate:          6942.80 [Kbytes/sec] received
```

## cargo bench

Version with Pin:
```
get_body_async_burst    time:   [118.97 us 120.46 us 121.87 us]
```

Version from Master:
```
get_body_async_burst    time:   [121.38 us 122.17 us 122.90 us]
```

**Comment by JohnTitor** [comment]
Rough reviews at this point:
- Do not update actix-{identify, session} here, we're moving out from here now, see https://github.com/actix/actix-extras.
- I'm negative to update all workspace members at once, we should do step by step _as possible_.
- The version numbers should be alpha since we may add breaking changes later. And I'm also negative to bump them here.

At a glance, the benchmark you commented looks good to me, so the change should be fine _roughly_.

**Comment by dunnock** [comment]
@JohnTitor need your advice. I initially wanted to include only changes in actix-http and point rest of crates to version 1.0.1. Though it appeared problematic as there is dependency chain:
actix-http  dev dependency is actix-http-test { ./test-server } which depends on awc, actix-web and actix-http
In turn awc depends on actix-http, actix-web and actix-http-test
actix-web depends on actix-http too


To illustrate the issue I have just pushed commit pointing actix-http to actix-http-test:1.0.0:

Problem is actix-http-test crate is used only to run some tests within actix-http crate, though if I do not provide updated version of awc and actix-web actix-http tests do not compile as they use different versions of structs and traits : actix-http:1.0.1 and actix-http:2.0.0. 

Once I update actix-web, crates actix-{identify, session} fail to compile.

One possible solution is to move out actix-http tests which depend on actix-http-test crate. Another option would be to comment out those tests temporarily.

**Comment by JohnTitor** [comment]
One safe way is to prepare another branch and work there. You can ignore workspace members that don't compile there. I just prepared `msg-body` branch.
And we can ignore actix-{identity, session} once the alpha version is released and they're published on extras repo.

**Comment by dunnock** [comment]
This is ready for review now with all 3 changes as requested. 
I believe CI issues not related to this change.

File: Cargo.toml
Comment (cdbattags) [inline Cargo.toml]: no extra whitespace plz

File: actix-http/src/encoding/encoder.rs
Comment (cdbattags) [inline actix-http/src/encoding/encoder.rs:126]: @robjtede are we ok with naming `this`?

File: actix-http/src/h1/dispatcher.rs
Comment (cdbattags) [inline actix-http/src/h1/dispatcher.rs]: Spacing here please between the `+`

**Review by cdbattags — COMMENTED** [review]
Small things but otherwise looks great

File: Cargo.toml
Comment (dunnock) [reply Cargo.toml]: Fixed

File: actix-http/src/h1/dispatcher.rs
Comment (dunnock) [reply actix-http/src/h1/dispatcher.rs]: Good point. I actually removed Unpins since MessageBody subtrait of Unpin now..

**Review by cdbattags — APPROVED** [review]

File: actix-http/src/encoding/encoder.rs
Comment (robjtede) [reply actix-http/src/encoding/encoder.rs:126]: im happy with it, seen it around and think its standard-ish practice

**Comment by robjtede** [comment]
Been thinkng about this for a couple days. It seems uneccesary to have MessageBody take the Unpin supertrait. You're handling the self Pin signatures correctly in most places anyway but it only takes a couple of simple modifications to make it work without the supertrait.

Eg. the `Box<T>` and `Box<dyn MessageBody>` impls can use `map_unchecked_mut` calls. The h1 dispatcher maybe slightly more involved (just from changing a couple things) but given the h2 dispatcher is already using pin_project it seems achievable.

**Review by robjtede — CHANGES_REQUESTED** [review]
Been thinkng about this for a couple days. It seems uneccesary to have MessageBody take the Unpin supertrait. You're handling the self Pin signatures correctly in most places anyway but it'll only takes a couple of modifications to make it work without the supertrait.

Eg. the Box<T> and Box<dyn MessageBody> impls can use map_unchecked_mut calls. The h1 dispatcher maybe slightly more involved (just from changing a couple things) but given the h2 dispatcher is already using pin_project it seems achievable.

**Comment by dunnock** [comment]
@robjtede I have unlinked MessageBody from Unpin as you suggested. h1::dispatcher appeared quite complicated and requires strong review. DispatcherState and SendResponse required some unsafe code too.

**Comment by Dowwie** [comment]
I thought everyone was in agreement not to introduce new unsafe to this project?  I really don't want to be the unsafe police here.

**Comment by cetra3** [comment]
Yep, I agree with @Dowwie that we should not introduce more `unsafe` if possible.

I have a feeling that [this line is unsound](https://github.com/actix/actix-web/pull/1332/files#diff-2a76b8092149c9558280c938b3622598R163) if Self does not implement Unpin.  Can `pin_project` be used here?

**Comment by dunnock** [comment]
I think we should decide here - without Unpin for MessageBody it seems inevitable to have unsafe. I have tried few implementations trying to skip unsafe, the only working was with Unpin as supertrait. But maybe bigger refactoring might help.

**Comment by cetra3** [comment]
I think that it makes sense to use `Unpin` unless there is a very good reason not to.  It's unsafe as you need to guarantee that what you are Pinning won't move otherwise.

I think the majority of types implement it automatically if I'm not mistaken?

**Comment by dunnock** [comment]
> I have a feeling that [this line is unsound](https://github.com/actix/actix-web/pull/1332/files#diff-2a76b8092149c9558280c938b3622598R163) if Self does not implement Unpin. Can `pin_project` be used here?

Indeed it might be for some self referrential type. But the problem is this line actually:
https://github.com/actix/actix-web/pull/1332/files#diff-2a76b8092149c9558280c938b3622598R798
So we have Pin<Dispatcher> there and by upgrade logic we shall transfer ownership of self.inner parts into FramedParts (which does not accept Pinned references).

So it's either even bigger refactoring or return Unpin to MessageBody.

**Comment by dunnock** [comment]
So I will wait for final call and either return Unpin or continue with refactoring..

**Comment by dunnock** [comment]
Hi @JohnTitor , will need some decision here, there are 2 implementations:
1. Up to commit https://github.com/actix/actix-web/pull/1332/commits/99b72a6bda1a924dd1c7f393ae67b145e64c7b31 MessageBody trait requires Unpin and there are no unsafe code
2. After changes in recent commits requested in message https://github.com/actix/actix-web/pull/1332#pullrequestreview-351999954 there is some seems unavoidable new unsafe code appeared, which seems not inline with future policy (by recent comments above)

Should I continue working on this PR down the unsafe code path or rollback latest changes up to working solution with Unpin supertrait?

**Comment by Aaron1011** [comment]
@dunnock: I'm working on changes to `pin-project` that will help with the line that you mentioned. I can open a PR to address it once this is merged.

**Comment by dunnock** [comment]
@Aaron1011: If you mean https://github.com/taiki-e/pin-project/pull/181 not sure how that will help here? [FramedParts takes ownership](https://github.com/actix/actix-web/pull/1332/files#diff-2a76b8092149c9558280c938b3622598R798) of a stream, hence we are taking the `Dispatcher::inner` value but it is a Pin.

**Comment by Aaron1011** [comment]
@dunnock: This is different - this idea is to have a `pinned_replace` method that will allow you to call `mem::replace` on a pinned enum, if the enum is currently set to a variant that has no `#[pin]` annotations.

I haven't yet opened a PR for it, but I will soon.

**Comment by dunnock** [comment]
@Aaron1011 in our case current enum variant is Pinned (`DispatcherState::Normal`), though we can confidently say that it's Pin not going to be called or accessed from anywhere as previous projections lifetime has ended on a line before.

**Comment by dunnock** [comment]
Last commits getting rid of unsafe code in a dispatcher with Aaron's idea to transfer not pinned fields (DispatchInner::io, *_buf, codec). Note that `unwrap()` should behave like a previous implementation which had explicit `panic!` under same unreachable conditions. @cetra3 @robjtede I guess it now addresses both concerns and can be reviewed again.

Compared to current master this PR removes 4 unsafe blocks from dispatcher not adding any new.

By benchmark it is still have no noticeable difference to the version in master:
```
> ab -c 40 -n 1000000 -k http://127.0.0.1:59993/
Concurrency Level:      40
Time taken for tests:   30.182 seconds
Complete requests:      1000000
Failed requests:        0
Keep-Alive requests:    1000000
Total transferred:      220000000 bytes
HTML transferred:       119000000 bytes
Requests per second:    33132.75 [#/sec] (mean)
Time per request:       1.207 [ms] (mean)
Time per request:       0.030 [ms] (mean, across all concurrent requests)
Transfer rate:          7118.36 [Kbytes/sec] received
```

**Comment by robjtede** [comment]
I'm liking the look of these changes @dunnock, thanks for putting the time in to this. Will give it a proper look over later today and resubmit review.

**Review by robjtede — APPROVED** [review]

**Comment by robjtede** [comment]
Really good work; the tendrils of that trait ran much deeper than I thought. I'm glad we kept the Unpin only to necessary parts.

**Comment by cdbattags** [comment]
@dunnock amazing work on this

thanks for the time

**Comment by Dowwie** [comment]
@JohnTitor do you see any reason not to merge?  would rather you took the lead here..

File: actix-http/Cargo.toml
Comment (JohnTitor) [inline actix-http/Cargo.toml]: Could you not bump up version here? It may cause conflict when merging (into) master.

**Review by JohnTitor — APPROVED** [review]
Really great as a first step! I left a comment for a nit point.

File: actix-http/Cargo.toml
Comment (Dowwie) [reply actix-http/Cargo.toml]: @dunnock

Comment (dunnock) [reply actix-http/Cargo.toml]: Reverted. I also temporarily excluded local actix-http resolution from actix-web so that it compiles, next step would be to adopt changes to actix-web and awc

Comment (dunnock) [reply actix-http/Cargo.toml]: Merged in last updates from master, fixed tests to account for actix-http is not compatible with other workspace packages now

**Review by JohnTitor — APPROVED** [review]
Sorry for the delay! Let's go ahead :)

**Pull request merged**
